go-controller/pkg/ovn/ovn.go

```diff
@@ -624,6 +624,19 @@ func (oc *Controller) WatchPods() {
 		UpdateFunc: func(old, newer interface{}) {
 			oldPod := old.(*kapi.Pod)
 			pod := newer.(*kapi.Pod)
+			// there may be a situation where this update event is not the latest
+			// and we rely on annotations to determine the pod mac/ifaddr
+			// this would create a situation where
+			// 1. addLogicalPort is executing with an older pod annotation, skips setting a new annotation
+			// 2. creates OVN logical port with old pod annotation value
+			// 3. CNI flows check fails and pod annotation does not match what is in OVN
+			// Therefore we need to get the latest version of this pod to attempt to addLogicalPort with
+			pod, err := oc.watchFactory.GetPod(pod.Namespace, pod.Name)
```

Comment:
we are clobbering the `newer` podspec with the latest in the cache. Will we leak some events as a result? Would it make sense rather to compare the annotations with the addresses on the LSP in the ovn nbdb cache?

Reply:
In this case, the nbdb cache has no entry for this LSP beacuse setting annotation worked, but OVN execute failed.
 
Subsequent update events will not run addLogicalPort as long as it returned success because it wont be in the retry cache (so addPort will be false):
```
	if util.PodWantsNetwork(pod) && addPort {
		if err := oc.addLogicalPort(pod); err != nil {
			klog.Errorf(err.Error())
			oc.recordPodEvent(err, pod)
			return false
		}
	} else {
		// either pod is host-networked or its an update for a normal pod (addPort=false case)
		if oldPod == nil || exGatewayAnnotationsChanged(oldPod, pod) || networkStatusAnnotationsChanged(oldPod, pod) {
			if err := oc.addPodExternalGW(pod); err != nil {
				klog.Errorf(err.Error())
				oc.recordPodEvent(err, pod)
				return false
			}
		}
	}
```

However, the else statement will still execute to check if exgw annotations changed. But since we are always grabbing latest pod, we dont really care about intermediary states (more level triggered). I think the end result should be less churn in updates.

Reply:
Ok makes sense. I think earlier we used to annotate at the very end? So perhaps we could compare with the cache.